pulp_container/app/registry_api.py

```diff
@@ -475,6 +480,134 @@ def get_queryset(self, *args, **kwargs):
         return (public_repositories | accessible_repositories).distinct()
 
 
+class FlatpakIndexDynamicView(APIView):
+    """
+    Handles requests to the /index/dynamic endpoint
+    """
+
+    authentication_classes = []
+    permission_classes = []
+
+    FLATPAK_LABEL_KEY = "org.flatpak.ref"
+
+    def get(self, request):
+        req_repositories = None
+        req_tags = None
+        req_oss = None
+        req_architectures = None
+        req_flatpak_label_exists = False
+        req_label_exists = set()
+        req_label_values = {}
+        for key, values in request.query_params.lists():
+            if key == "repository":
+                req_repositories = values
+            elif key == "tag":
+                req_tags = values
+            elif key == "os":
+                req_oss = values
+            elif key == "architecture":
+                req_architectures = values
+            elif key.startswith("label:"):
+                if key.endswith(":exists"):
+                    if any(v != "1" for v in values):
+                        raise ParseError()
+                    label = key[len("label:") : len(key) - len(":exists")]
+                    if label == self.FLATPAK_LABEL_KEY:
+                        req_flatpak_label_exists = True
+                    else:
+                        req_label_exists.add(label)
+                else:
+                    label = key[len("label:") :]
+                    req_label_values[label] = values
+            else:
+                # In particularly, this covers any annotation:... parameters, which this
+                # implementation does not support:
+                raise ParseError()
+
+        if not req_flatpak_label_exists:
+            raise ParseError()
+
+        distributions = models.ContainerDistribution.objects.filter(private=False).only("base_path")
+
+        if req_repositories:
+            distributions = distributions.filter(base_path__in=req_repositories)
+
+        results = []
+        for distribution in distributions:
+            images = []
+            if distribution.repository:
+                repository_version = distribution.repository.latest_version()
+            elif distribution.repository_version:
+                repository_version = distribution.repository_version
+            tags = models.Tag.objects.filter(pk__in=repository_version.content)
+            manifests = models.Manifest.objects.filter(
+                pk__in=repository_version.content,
+                media_type__in=(models.MEDIA_TYPE.MANIFEST_V2, models.MEDIA_TYPE.MANIFEST_OCI),
+            )
```

Comment:
I think this is the right approach even if if we are trying to get an accurate Tags list for the index - I'd first find the matching images, *then* figure out the Tags that apply to the image. But actually, I would suggest not providing an accurate Tags, if it requires extra database queries and work. Flatpak doesn't care about the value of Tags, and it's only a SHOULD in the spec.
  
> Tags: List of all tags applied to the image. This appears only for images appearing in the Images list of a repository, and not for images in an image List. All tags applied to the image SHOULD be returned, even if a tag parameter in the query limits the images returned by tag.

Reply:
> I would note here that a query from a Flatpak client will _always_ include a tag= query parameter - and this can be the biggest limiter on the resulting content set.
> 
> My instinct, from a performance point of view, would be to accumulate a set of manifests that are tagged with the tags in the query (directly or indirectly), then only parse _those_ manifests to apply further filters such as label, os, architecture.
> 
> This could easily result in factors of 100 less config blobs being parsed, since it wouldn't be shocking to have hundreds of historical tags.

done with <https://github.com/pulp/pulp_container/pull/1301/commits/bd60171eb85a95d3a0f70d23dadea97ea334da14> "Optimize for common case of filtering on requested tags"

> The request will also always include an architecture. I think it would be fine to apply the os/architecture filters when querying for manifests tagged by manifest list, despite the wording in the spec that the os/architecture come from the config.json.
> 
> This would reduce parsing config blobs by up to a factor of 4 - if someone was literally mirroring manifest lists with all architectures from registry.redhat.io - I don't know if there's any provision for subsetting architectures when mirroring.

done with <https://github.com/pulp/pulp_container/pull/1301/commits/9c9ac62758ff19a15afbea2ecfd53a42530133f3> "Also filter on os/arch in recurse_through_manifest_lists"

> I think this is the right approach even if if we are trying to get an accurate Tags list for the index - I'd first find the matching images, _then_ figure out the Tags that apply to the image. But actually, I would suggest not providing an accurate Tags, if it requires extra database queries and work. Flatpak doesn't care about the value of Tags, and it's only a SHOULD in the spec.

done with <https://github.com/pulp/pulp_container/pull/1301/commits/0942e37ad485509bf28a51ad75f4e145b528dd82> "To improve performance, only report requested tags"